test/src/test-cppapi-consolidation-plan.cc

```diff
@@ -213,6 +215,225 @@ TEST_CASE_METHOD(
+  auto uri1 = write_sparse(a1, d, d, 1);
+  auto uri2 = write_sparse({0, 1}, {2, 4}, {2, 4}, 2);
+
+  // Create a consolidation plan with max fragment size of 10k. That way, only
+  // the first fragment needs to be split.
+  Array array{ctx_, SPARSE_ARRAY_NAME, TILEDB_READ};
+  ConsolidationPlan consolidation_plan(ctx_, array, 10 * 1024);
+
+  // Validate the plan.
+  CHECK(consolidation_plan.num_nodes() == 1);
+  CHECK(consolidation_plan.num_fragments(0) == 1);
+  CHECK(consolidation_plan.fragment_uri(0, 0) == uri1);
+}
+
+TEST_CASE_METHOD(
+    CppConsolidationPlanFx,
+    "C++ API: Consolidation plan, combine small 1",
+    "[cppapi][consolidation_plan][combine-small-1]") {
+  create_sparse_array(true);
+
+  // Write one fragment with NED [1, 2][1, 2] and one with NED [3, 4][3, 4].
+  auto uri1 = write_sparse({0, 1}, {1, 2}, {1, 2}, 1);
+  auto uri2 = write_sparse({2, 3}, {3, 4}, {3, 4}, 2);
+
+  // Create a consolidation plan with max fragment size of 100k. That way, the
+  // smaller fragments are considered for combining.
+  Array array{ctx_, SPARSE_ARRAY_NAME, TILEDB_READ};
+  ConsolidationPlan consolidation_plan(ctx_, array, 100 * 1024);
+
+  // Validate the plan.
+  CHECK(consolidation_plan.num_nodes() == 1);
+  CHECK(consolidation_plan.num_fragments(0) == 2);
+  CHECK(consolidation_plan.fragment_uri(0, 0) == uri1);
+  CHECK(consolidation_plan.fragment_uri(0, 1) == uri2);
+}
+
+TEST_CASE_METHOD(
+    CppConsolidationPlanFx,
+    "C++ API: Consolidation plan, combine small 2",
+    "[cppapi][consolidation_plan][combine-small-2]") {
+  create_sparse_array(true);
+
+  // Write one fragment with NED [1, 2][1, 2] and one with NED [5, 6][5, 6]. In
+  // the middle at [3, 3], we add a large fragment which will prevent
+  // consolidation.
+  auto uri1 = write_sparse({0, 1}, {1, 2}, {1, 2}, 1);
+
+  std::vector<int> a1(10 * 1024);
+  std::vector<uint64_t> d(10 * 1024);
+  std::fill(a1.begin(), a1.end(), 2);
+  std::fill(d.begin(), d.end(), 3);
+  auto uri2 = write_sparse(a1, d, d, 2);
+
+  auto uri3 = write_sparse({3, 4}, {5, 6}, {5, 6}, 3);
+
+  // Create a consolidation plan with max fragment size of 100k. That way, the
+  // smaller fragments are considered for combining.
+  Array array{ctx_, SPARSE_ARRAY_NAME, TILEDB_READ};
+  ConsolidationPlan consolidation_plan(ctx_, array, 100 * 1024);
+
+  // Validate the plan.
+  CHECK(consolidation_plan.num_nodes() == 1);
+  CHECK(consolidation_plan.num_fragments(0) == 1);
+  CHECK(consolidation_plan.fragment_uri(0, 0) == uri2);
+}
+
+TEST_CASE_METHOD(
+    CppConsolidationPlanFx,
+    "C++ API: Consolidation plan, complex 1",
+    "[cppapi][consolidation_plan][complex 1]") {
+  create_sparse_array(true);
+
+  // Write one fragment with NED [11, 14][11, 14] and one with NED [14, 16][14,
+  // 16]. Since they intersect, they should be batched together.
+  auto uri1 = write_sparse({0, 1}, {11, 14}, {11, 14}, 1);
+  auto uri2 = write_sparse({2, 3}, {14, 16}, {14, 16}, 2);
```

Comment:
I noticed that in all the tests we write the fragments that intersect the one after the other. It would be a good idea to write some non intersecting fragment in between just to have code coverage for that scenario as well. I think your algorithm will work fine but it will make the tests more robust to future changes.